### PR TITLE
fix query params for List PRs API call for stash provider for paginat…

### PR DESCRIPTION
**Pull request by karansaraswat19**
…ion and state of PRs

### CODE REVIEW

File: scm/pr.go [hunk 1]
```diff
@@ -48,6 +48,7 @@ type (
 		Size   int
 		Open   bool
 		Closed bool
+		Merged bool
```
Comment (abhinav-harness) [inline scm/pr.go:51]: this is added in a common file, we would need addition for all providers for this if they support

File: scm/driver/stash/util.go [hunk 2]
```diff
@@ -108,10 +108,12 @@ func encodePullRequestListOptions(opts scm.PullRequestListOptions) string {
 	if opts.Size != 0 {
 		params.Set("limit", strconv.Itoa(opts.Size))
 	}
-	if opts.Open && opts.Closed {
+	if opts.Open && opts.Closed && opts.Merged {
```
Comment (abhinav-harness) [inline scm/driver/stash/util.go:111]: of what i understand we need state all -> lets not make too many changes and in caller add both open and close

File: scm/pr.go (hunk 1 above)
Comment (karansaraswat19) [reply scm/pr.go:51]: Gitlab now has 4 states -> `opened, closed, locked, merged`
Bitbucket also now has 4 -> `OPEN, MERGED, DECLINED, SUPERSEDED`
Till now everywhere we have this condition though
`if opts.Open && opts.Closed {
	params.Set("state", "all")
}`

File: scm/driver/stash/util.go (hunk 2 above)
Comment (karansaraswat19) [reply scm/driver/stash/util.go:111]: did this according to the changes already present with this condition
`if opts.Open && opts.Closed { params.Set("state", "all") }`

File: scm/pr.go (hunk 1 above)
Comment (karansaraswat19) [reply scm/pr.go:51]: this might affect the usage of this method

**Review by bradrydzewski — CHANGES_REQUESTED** [review]
When adding new functionality to this library at the abstraction layer (in this case `scm.PullRequest.Merged`) we need to add support for multiple vendors (e.g. GitHub, GItLab and Stash) to ensure the abstraction is correct.

File: scm/pr.go (hunk 1 above)
Comment (bradrydzewski) [reply scm/pr.go:51]: one requirement to merging new functionality in this library is that it needs to be implemented for 3 or more providers to ensure the abstractions are correct.  The differences you highlighted are the reason this is required.

Comment (karansaraswat19) [reply scm/pr.go]: removed

File: scm/driver/stash/util.go (hunk 2 above)
Comment (karansaraswat19) [reply scm/driver/stash/util.go]: fixed

**Comment by karansaraswat19** [comment]
> When adding new functionality to this library at the abstraction layer (in this case `scm.PullRequest.Merged`) we need to add support for multiple vendors (e.g. GitHub, GItLab and Stash) to ensure the abstraction is correct.

Noted

**Review by abhinav-harness — APPROVED** [review]

File: scm/driver/stash/util.go [hunk 3]
```diff
@@ -111,7 +111,7 @@ func encodePullRequestListOptions(opts scm.PullRequestListOptions) string {
 	if opts.Open && opts.Closed {
 		params.Set("state", "all")
 	} else if opts.Closed {
-		params.Set("state", "closed")
+		params.Set("state", "declined")
```
Comment (karansaraswat19) [inline scm/driver/stash/util.go:114]: stash calls its closed PRs as "declined" PRs
```
"errors": [
        {
            "context": null,
            "message": "Pull request state should be one of: [DECLINED, MERGED, OPEN]",
            "exceptionName": null
        }
    ]
    ```

**Pull request merged**